kube/client.go

```diff
@@ -148,41 +160,183 @@ func isCompatible(clientMajor, clientMinor, serverMajor, serverMinor string) err
+
+	tempKubeConfigFilepath, tempCertFilepath, err := c.CreateTempConfig(namespace, "kube-applier")
+	if err != nil {
+		return "", "", fmt.Errorf("error creating temp config: %v", err)
+	}
+	defer func() { os.Remove(tempKubeConfigFilepath); os.Remove(tempCertFilepath) }()
+	args = append(args, fmt.Sprintf("--kubeconfig=%s", tempKubeConfigFilepath))
+
+	return executeApply(args)
+}
+
+func decodeToJson(i interface{}, raw_message []byte) error {
+	dec := json.NewDecoder(bytes.NewReader(raw_message))
+
+	if err := dec.Decode(i); err != nil {
+		return fmt.Errorf("Response is not json format: %v error=(%v)", string(raw_message), err)
+	}
+
+	return nil
+}
+
 func (c *Client) GetNamespaceStatus(namespace string) (AutomaticDeploymentOption, error) {
 	args := []string{"kubectl", "get", "namespace", namespace, "-o", "json", "-n", namespace}
 	if c.Server != "" {
 		args = append(args, fmt.Sprintf("--kubeconfig=%s", kubeconfigFilePath))
 	}
-	stdout, err := exec.Command(args[0], args[1:]...).CombinedOutput()
+	stdout, err := execCommand(args[0], args[1:]...).CombinedOutput()
 	if err != nil {
 		return Off, err
 	}
-	dec := json.NewDecoder(bytes.NewReader(stdout))
+
 	var nr struct {
 		Metadata struct {
 			Labels map[string]string
 		}
 	}
-	if err := dec.Decode(&nr); err != nil {
-		return Off, fmt.Errorf("Get namespace response is not json format: %v error=(%v)", string(stdout), err)
+	if err := decodeToJson(&nr, stdout); err != nil {
+		return Off, err
 	}
+
 	return AutomaticDeploymentOption(nr.Metadata.Labels[c.Label]), nil
 }
+
+func (c *Client) GetNamespaceUserSecretName(namespace, username string) (string, error) {
+	args := []string{"kubectl", "get", "serviceaccount", username, "-o", "json", "-n", namespace}
+	if c.Server != "" {
+		args = append(args, fmt.Sprintf("--kubeconfig=%s", kubeconfigFilePath))
+	}
+	stdout, err := execCommand(args[0], args[1:]...).CombinedOutput()
+	if err != nil {
+		return "", err
+	}
+
+	type secret struct {
+		Name string
+	}
+	var nr struct {
+		Secrets []secret
+	}
+	if err := decodeToJson(&nr, stdout); err != nil {
+		return "", err
+	}
+
+	if len(nr.Secrets) > 1 {
+		log.Printf("Found many secrets for kube-applier on %s using the first on the list", namespace)
+	}
+	if len(nr.Secrets) == 0 {
+		return "", fmt.Errorf("No secrets found for %s user", username)
+	}
+	return nr.Secrets[0].Name, nil
+}
+
+func (c *Client) GetUserDataFromSecret(namespace, secret string) (string, string, error) {
+	args := []string{"kubectl", "get", "secret", secret, "-o", "json", "-n", namespace}
+	if c.Server != "" {
+		args = append(args, fmt.Sprintf("--kubeconfig=%s", kubeconfigFilePath))
+	}
+	stdout, err := execCommand(args[0], args[1:]...).CombinedOutput()
+	if err != nil {
+		return "", "", err
+	}
+
+	var nr struct {
+		Data map[string]string
+	}
+	if err := decodeToJson(&nr, stdout); err != nil {
+		return "", "", err
+	}
+
+	return nr.Data["token"], nr.Data["ca.crt"], nil
```

Comment:
Shouldn't we check for existence of the keys? Like:
```
token, ok := nr.Data["token"]
if !ok{
   return "","", errors.New("..")
}
```

the app will panic if you don't do that

Reply:
Excellent point, thanks!